File: problems/0329_longest-increasing-path-in-a-matrix/solution.go
import "sort"

func longestIncreasingPath(matrix [][]int) int {
	m, n := len(matrix), len(matrix[0])
	type cell struct{ v, i, j int }
	cells := make([]cell, 0, m*n)
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			cells = append(cells, cell{matrix[i][j], i, j})
		}
	}
	sort.Slice(cells, func(a, b int) bool {
		return cells[a].v < cells[b].v
	})
	dp := make([][]int, m)
	for i := range dp {
		dp[i] = make([]int, n)
		for j := range dp[i] {
			dp[i][j] = 1
		}
	}
	best := 1
	dirs := [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
	for _, c := range cells {
		for _, d := range dirs {
			ni, nj := c.i+d[0], c.j+d[1]
			if ni >= 0 && ni < m && nj >= 0 && nj < n && matrix[ni][nj] < c.v {
				if dp[ni][nj]+1 > dp[c.i][c.j] {
					dp[c.i][c.j] = dp[ni][nj] + 1
				}
			}
		}
		if dp[c.i][c.j] > best {
			best = dp[c.i][c.j]
		}
	}
	return best
}

File: problems/0329_longest-increasing-path-in-a-matrix/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    public int longestIncreasingPath(int[][] matrix) {
        int m = matrix.length,
            n = matrix[0].length;
        List<int[]> cells = new ArrayList<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) cells.add(new int[] {
                matrix[i][j],
                i,
                j,
            });
        }
        cells.sort((a, b) -> Integer.compare(a[0], b[0]));
        int[][] dp = new int[m][n];
        for (int i = 0; i < m; i++) for (int j = 0; j < n; j++) dp[i][j] = 1;
        int best = 1;
        int[][] dirs = { { 1, 0 }, { -1, 0 }, { 0, 1 }, { 0, -1 } };
        for (int[] cell : cells) {
            int v = cell[0],
                i = cell[1],
                j = cell[2];
            for (int[] d : dirs) {
                int ni = i + d[0],
                    nj = j + d[1];
                if (
                    ni >= 0 && ni < m && nj >= 0 && nj < n && matrix[ni][nj] < v
                ) {
                    if (dp[ni][nj] + 1 > dp[i][j]) dp[i][j] = dp[ni][nj] + 1;
                }
            }
            if (dp[i][j] > best) best = dp[i][j];
        }
        return best;
    }
}
